### PR TITLE
Migrate scripts to python 3

### LINKED ISSUE #5950: Refactor: Update our python scripts to Python 3
Our current python scripts and test frameworks work only under Python 2.7; there are a few changes that we need to make to upgrade our scripts to python 3, notably concerning encoding of strings. But toolings exist to helps us update and compatibility layers also exist to ease the transition.

I understand that python 2.7 EOL is in [2020](http://legacy.python.org/dev/peps/pep-0373/), but we should still work on that.

related to a discuss post: https://discuss.elastic.co/t/step-3-init-and-create-the-metricset-failure/112046/5

### PR DESCRIPTION
**Pull request by jsoriano**
PR to identify and fix issues with python 3.

**If you want to help**:
* Pick up an issue from the list of issues identified
* If you have identified an issue that is not in the list, add it there
* Add your name next to the item in the list (if you cannot edit this description, add a comment)
* When you have a fix, open a PR, with `feature/python3` as base branch
* Once the issue is solved and the branch merged, check it in the list

**How to use Python 3 locally**:

If Python 3 is not the default in your system you may need to:
* Remove existing virtual environments you have for beats (you can selectively remove `build/python-env` and `build/ve` directories, or just run `make clean`)
* Install Python 3 in your system if it was not already installed

**Issues**:

Log of changes that we may consider to revert before merging:

- [x] Forced use of python3 using `PYTHON_EXE` env var
- [x] Don't check vendor and build directories on `make check` from the root directory
- [x] Move `check` to the `test` stage in Travis so failures there don't block other jobs so we can see their problems.

Log of issues identified/done:

- [x] Remove functools2 library from `requirements.txt`
- [x] Apply recommendations of `2to3` tool
- [x] Apply [recommendations](https://python-future.org/automatic_conversion.html) from `futurize` 
- [x] Don't reuse python virtualenvs from the host when running tests in docker https://github.com/elastic/beats/pull/14903 @jsoriano 
- [x] Python 2 compatibility shims and casts removed https://github.com/elastic/beats/pull/14989 @beniwohli 
- [x] Use `python3` in the shebang for python scripts. [`python` is going to be reserved for `python2` in Ubuntu](https://wiki.ubuntu.com/Python/3), and [PEP394](https://legacy.python.org/dev/peps/pep-0394/#abstract) mentions that _`python` should be used in the shebang line only for scripts that are source compatible with both Python 2 and 3_ https://github.com/elastic/beats/pull/15701
- [x] Review docs to look for pending references to Python 2 @ChrsMark  https://github.com/elastic/beats/pull/15702
- [x] Replace `virtualenv` with `python3 -m venv` and get rid of `PYTHON_EXE`. https://github.com/elastic/beats/pull/15754
- [x] Builds are slower in Travis, hitting timeouts in some cases. #15931
- [x] `make check` https://github.com/elastic/beats/pull/15226 @mikemadden42
  - [x] `dev-tools/generate_notice.py`:
    - [x] Changed byte array to string in `dev-tools/generate_notice.py` @jsoriano 
    - [x] Something related to encoding fails in CI
    - [x] Alternative to solve issues in `dev-tools/generate_notice.py`: migrate it to mage
  - [x] Once green, revert change in travis so `check` is run in a previous `check` stage #15817
  - [x] Different order in `fields.asciidoc` makes `make check` to randomly fail on Jenkins #15898 @jsoriano
    > -example: {'application': 'foo-bar', 'env': 'production'}
    > +example: {'env': 'production', 'application': 'foo-bar'}
- [x] `dev-tools/cherrypick_pr` #14900 #14929 @kvch 
- [x] Fix versions of python in CI
  - [x] Install Python 3 in jenkins workers https://github.com/elastic/infra/pull/16269 @jsoriano
  - [x] Upgrade Python 3 (or ubuntu base image) in Travis, it is using 3.4 that is discontinued https://github.com/elastic/beats/pull/15226 @mikemadden42
  - [x] Python 3 venv not available in Ubuntu CIs in beats-ci @jsoriano 
  - [x] Windows jenkins workers are using python 2 #16141
  - [x] ~~Consider keeping support also with Python 2 if some CI environment cannot be upgraded~~
 - [x] Update Dockerfiles used for running beats during testing to have Python3 https://github.com/elastic/beats/pull/15754
- [x] libbeat testsuite
  - [x] `Failed to connect: x509: certificate is valid for 127.0.0.1, not 127.0.1.1` (probably similar to https://github.com/elastic/beats/pull/15583) #15591 @jsoriano 
  - [x] There is still [something](https://travis-ci.org/elastic/beats/jobs/637569435#L1780) running on python 2.7 #15608 @jsoriano 
  - [x] Failures in test_management.py (see [this build](https://travis-ci.org/elastic/beats/jobs/637897077#L1098)) https://github.com/elastic/beats/pull/15728
- [x] Filebeat testsuite @kvch 
  - [x] x-pack suite is still being tested on Python 2 #14912 @jsoriano 
  - [x] `AssertionError: '0o640' != '0640'` https://github.com/elastic/beats/pull/15030
  - [x] Type errors (`'bool' object is not iterable`, `a bytes-like object is required, not 'str'`, `'float' object cannot be interpreted as an integer`). https://github.com/elastic/beats/pull/15030
  - [x] `ValueError: can't have unbuffered text I/O` https://github.com/elastic/beats/pull/15030
  - [x] `AttributeError: 'str' object has no attribute 'decode'` https://github.com/elastic/beats/pull/15030
- [x] Heartbeat testsuite
  - [x] `ImportError: No module named http.server` https://github.com/elastic/beats/pull/15111 @kvch
  - [x] Local IP not the same (after upgrading ubuntu image?) `objects not equal: actual(string(127.0.1.1)) != expected(string(127.0.0.1))` https://github.com/elastic/beats/pull/15583 @jsoriano 
- [x] Auditbeat testsuite
  - [x] Type errors (`a bytes-like object is required, not 'str'`) https://github.com/elastic/beats/pull/15111 @kvch
  - [x] [`test connected UDP IPv4 flow`](https://travis-ci.org/elastic/beats/jobs/639438111#L925)(needs cherry-pick to branch https://github.com/elastic/beats/pull/15771) fix backport as part of #15807
  - [x] Tests were skipped on Linux https://github.com/elastic/beats/pull/15727
- [x] Metricbeat testsuite
  - [x] `ImportError: No module named cli.command` (`from compose.cli.command import get_project`)
  - [x] `ImportError: No module named request` (`import urllib.request`, was autogenerated by `2to3`)
  - [x] `AttributeError: 'Test' object has no attribute 'assertNotRegexpMatches'` https://github.com/elastic/beats/pull/15111
  - [x] ActiveMQ test failures and dashbards with incorrect versions #15807
  - [x] Missing skip for ES integration tests #15779
  - [x] Appsearch tests fail #15878 :crossed_fingers:  @jsoriano 
- [x] Packetbeat testsuite
  - [x] `TypeError: Unicode-objects must be encoded before hashing`
- [x] Functionbeat testsuite
  - [x] `TypeError: 'str' does not support the buffer interface` #14901 @kvch
- [x] Winlogbeat pythonUnitTest
  - [x] Fix unicode issues https://github.com/elastic/beats/pull/15831
- [x] `make docs` - to be removed @jsoriano https://github.com/elastic/beats/pull/15816
  - [x] ~~Docker build [fails in travis](https://travis-ci.org/elastic/beats/jobs/638355471#L467)~~
- [x] vagrant file https://github.com/elastic/beats/blob/b74e19cd1ff873c3c8d111bcf14403ca5f91829c/Vagrantfile#L85
  - [x] Test Windows 2019 Vagrantbox with Python3 https://github.com/elastic/beats/pull/15831

**To do when build is green, before merging:**

 - [x] Prepare migration guidelines for contributors working with Python 2 @jsoriano #15833
 - [x] Sync with APM Server to decide when to merge @jsoriano 

Fixes #5950.

### CODE REVIEW

**Comment by kvch** [comment]
Opened small PR: https://github.com/elastic/beats/pull/14900

**Comment by jsoriano** [comment]
jenkins, test this again to see if python 3 is already available in workers

**Comment by beniwohli** [comment]
Great stuff, fantastic to see some more Python 3 adoption!

Is the plan to switch over from Python 2 to Python 3, or to support both at the same time? If it's the former, there's a lot of changes in the PR that aren't really necessary (I assume those were introduced by `2to3` or `futurize`). Some examples:

 * all `from __future__` imports
 * all `from builtins` imports
 * `class SomeClass:` -> `class SomeClass(object)` (in Python 3, all classes inherit from `object`, there aren't old-style classes anymore)
 * many of the the instances where an iterator is cast to a list, only to iterate over it, e.g.
      
        for _, deps in list(dependencies.items()):
 * 
        from future import standard_library
        standard_library.install_aliases()

       This installs aliases in Python 2 for modules that moved around in Python 3, so not needed if you target Python 3 only

While these changes don't really hurt, it makes the diff quite a bit larger than it could be, and the code itself is not quite as readable.

If it helps, I can open a PR that removes these unnecessary changes (again, assuming you don't plan to support running the tests with Python 2 going forward).

**Comment by jsoriano** [comment]
Hey @beniwohli, thanks for your comment!

Yes, most of this code was added by `futurize`, and I thought in principle that it could be a good idea to support both versions, but after a conversation with @kvch I saw that supporting Python 2 can be an overhead and it gives very few benefits, specially for this code that is only used for tooling and testing. So in principle we are not going to support both versions. Any cleanup you can make about this (or any other thing you see) is more than welcome :slightly_smiling_face: 

I thought on keeping support Python 2 at least during the migration, so developers and contributors have some time to adapt their development environments, and we make sure that we can run Python 3 in all CI environments, but it is probably less effort to make a one-time migration of everything to Python 3 than supporting both versions.

**Comment by beniwohli** [comment]
As someone who has to maintain a dual-version code base, I highly recommend doing a clean cut from Python 2 to Python 3. While it is possible, you'll suffer a lot of pain. In this specific case, I think it totally makes sense to require people to install Python 3 if they want to run the test suite locally.

**Comment by andrewkroh** [comment]
I think we should also switch over to use `python3 -m venv` instead of `virtualenv` because it's part of Python3 and is the recommend way. https://docs.python.org/3/library/venv.html

**Comment by jsoriano** [comment]
> I think we should also switch over to use `python3 -m venv` instead of `virtualenv` because it's part of Python3 and is the recommend way. https://docs.python.org/3/library/venv.html
> 

This sounds great to me, but what do you think about doing that as a follow up step?

**Comment by jsoriano** [comment]
> > I think we should also switch over to use `python3 -m venv` instead of `virtualenv` because it's part of Python3 and is the recommend way. https://docs.python.org/3/library/venv.html
> 
> This sounds great to me, but what do you think about doing that as a follow up step?

After talking on slack we have decided that the change of `virtualenv` to `python3 -m venv` would be an easy one and would also allow to get rid of `PYTHON_EXE`, simplifying also the migration of existing workspaces. So we are going for it on this branch. Thanks @andrewkroh !

**Comment by adriansr** [comment]
The `test connected UDP IPv4 flow` failure has nothing to do with Python 3, is an actual failure in the system.socket dataset due to kernel 5.x.

**Comment by andrewkroh** [comment]
I'll go ahead and mark those failing auditbeat UDP tests as skipped for this for this branch.

**Comment by jsoriano** [comment]
jenkins, test this again please

**Comment by jsoriano** [comment]
jenkins, test this again please

**Comment by jsoriano** [comment]
jenkins, test this again please

**Comment by jsoriano** [comment]
After https://github.com/elastic/beats/pull/15878 I think that pending failures in travis are caused by some builds taking significative longer times. We will have to investigate this.

**Comment by jsoriano** [comment]
jenkins, test this again please

**Comment by jsoriano** [comment]
jenkins, test this again please

**Comment by mikemadden42** [comment]
run beats-ci/package

**Comment by jsoriano** [comment]
jenkins, test this again please

**Comment by jsoriano** [comment]
run beats-ci/package

**Comment by jsoriano** [comment]
@mikemadden42 packaging build is failing in OSX: https://beats-ci.elastic.co/job/elastic+beats+pull-request+package/31/

Should we fix it before merging or is is ok if we fix it later?

**Comment by andrewkroh** [comment]
Ignore the OSX packaging job. The OSX packaging build needs to be disabled/removed until there's a more powerful worker available. It doesn't have enough memory to complete the build.

**Comment by jsoriano** [comment]
jenkins, test this again please

**Comment by jsoriano** [comment]
Opened PR with the squashed changes to prepare the backport to 7.x https://github.com/elastic/beats/pull/16302

**Comment by urso** [comment]
x-pack/metricbeat did fail on travis on `go vet` running out of memory.

Checking beats-ci and travis-ci, it looks like each testsuite did succeed at least once.

File: x-pack/metricbeat/module/appsearch/docker-compose.yml [hunk 1]
```diff
@@ -2,8 +2,11 @@ version: '2.3'
 
 services:
   appsearch:
+    image: docker.elastic.co/integrations-ci/beats-appsearch:${APPSEARCH_VERSION:-7.5.0}-1
     build:
       context: ./_meta
+      args:
+        APPSEARCH_VERSION: ${APPSEARCH_VERSION:-7.5.0}
```
Comment (urso) [inline x-pack/metricbeat/module/appsearch/docker-compose.yml]: The default version should be 8.0.0-SNAPSHOT, so to test against the app-search master branch.

Comment (jsoriano) [reply x-pack/metricbeat/module/appsearch/docker-compose.yml]: Appsearch version is set to 7.5.0 also in master, we will review it as a follow up.

File: dev-tools/mage/pytest.go [hunk 2]
```diff
@@ -32,6 +32,14 @@ import (
 	"github.com/magefile/mage/sh"
 )
 
+// WINDOWS USERS:
+// The python installer does not create a python3 alias like it does on other
+// platforms. So do verify the version with python.exe --version.
+//
+// Setting up a python virtual environment on a network drive does not work
+// well. So if this applies to your development environment set PYTHON_EXE
```
Comment (andrewkroh) [inline dev-tools/mage/pytest.go]:
```suggestion
// well. So if this applies to your development environment set PYTHON_ENV
```

Comment (jsoriano) [reply dev-tools/mage/pytest.go]: Fixed in https://github.com/elastic/beats/pull/16364, where I am also adding a changelog entry for this.

**Pull request merged**